File: archive/j/javascript/convex-hull.js
/**
 * Calculate the Convex Hull of a set of 2D points using Graham's Scan algorithm.
 *
 * @param {Array} points - An array of 2D points represented as objects with x and y properties.
 * @returns {Array} - An array of points on the convex hull in counter-clockwise order.
 */
function convexHull(points) {
	function orientation(p, q, r) {
		const val = (q.y - p.y) * (r.x - q.x) - (q.x - p.x) * (r.y - q.y)
		if (val === 0) return 0
		return val > 0 ? 1 : -1
	}

	const n = points.length
	if (n < 3) return points

	let minPointIndex = 0
	for (let i = 0; i < n; i++) {
		if (
			(points[i].y < points[minPointIndex].y) ||
			(points[i].y == points[minPointIndex].y && points[i].x < points[minPointIndex].x)
		) {
			minPointIndex = i
		}
	}
	const sortedPoints = [...points].sort((a, b) => {
		const angleA = Math.atan2(a.y - points[minPointIndex].y, a.x - points[minPointIndex].x)
		const angleB = Math.atan2(b.y - points[minPointIndex].y, b.x - points[minPointIndex].x)
		return angleA - angleB
	})

	const convexHull = [sortedPoints[0], sortedPoints[1]]

	for (let i = 2; i < n; i++) {
		while (
			convexHull.length > 1 &&
			orientation(
				convexHull[convexHull.length - 2],
				convexHull[convexHull.length - 1],
				sortedPoints[i]
			) !== -1
		) {
			convexHull.pop()
		}
		convexHull.push(sortedPoints[i])
	}

	return convexHull
}

/**
 * Executable function for Convex Hull calculation with command-line input.
 *
 * @param {string} inputX - Comma-separated X coordinates as a string.
 * @param {string} inputY - Comma-separated Y coordinates as a string.
 */
function main(inputX, inputY) {
	if (inputX == undefined || inputY == undefined){
		inputX = " "
		inputY = " "
	}
	const xArray = inputX.split(",").map(Number)
	const yArray = inputY.split(",").map(Number)

	if (
		xArray.length < 3 ||
		yArray.length < 3 ||
		xArray.length !== yArray.length ||
		xArray.some(item => isNaN(item))|| yArray.some(item => isNaN(item))
	) {
		console.log(
			'Usage: please provide at least 3 x and y coordinates as separate lists (e.g. "100, 440, 210")'
		)
		return
	}

	const points = xArray.map((x, i) => ({ x, y: yArray[i] }))

	const convexHullResult = convexHull(points)
	convexHullResult.forEach((point) => console.log(`(${point.x}, ${point.y})`))
}

// Run the executable function with command-line arguments
const inputX = process.argv[2]
const inputY = process.argv[3]
main(inputX, inputY)

File: archive/j/java/ConvexHull.java
import java.util.*;

public class ConvexHull {
    public static void main(String[] args) {
        if (args.length < 2)
            System.out.println(
                    "Usage: please provide at least 3 x and y coordinates as separate lists (e.g. \"100, 440, 210\")");
        else {
            String xInput = args[0];
            String yInput = args[1];
            String[] tempX = xInput.split(", ");
            String[] tempY = yInput.split(", ");
            if (tempX.length != tempY.length || tempX.length < 3) {
                System.out.println(
                        "Usage: please provide at least 3 x and y coordinates as separate lists (e.g. \"100, 440, 210\")");
                return;
            }
            Point[] points = new Point[tempX.length];
            int minX = Integer.MAX_VALUE;
            int mindex = 0;
            for (int a = 0; a < tempX.length; a++) {
                if (!tempX[a].matches("-?\\d+") || !tempY[a].matches("-?\\d+")) {
                    System.out.println(
                            "Usage: please provide at least 3 x and y coordinates as separate lists (e.g. \"100, 440, 210\")");
                    return;
                }
                points[a] = new Point(Integer.parseInt(tempX[a]), Integer.parseInt(tempY[a]));
                if (points[a].x <= minX) {
                    minX = points[a].x;
                    mindex = a;
                }
            }
            Stack<Point> result = new Stack<Point>();
            int considerdex = mindex;
            do {
                result.push(points[considerdex]);
                int nextdex = (considerdex + 1) % points.length;
                for (int contenderdex = 0; contenderdex < points.length; contenderdex++) {
                    if (isCounterClockwise(points[considerdex], points[nextdex], points[contenderdex]))
                        nextdex = contenderdex;
                }
                considerdex = nextdex;
            } while (considerdex != mindex);
            int k = result.size();
            for (int a = 0; a < k; a++) {
                System.out.println(result.pop().toString());
            }
        }
    }

    public static boolean isCounterClockwise(Point a, Point b, Point c) {
        return (c.y - a.y) * (b.x - a.x) > (b.y - a.y) * (c.x - a.x);
    }
}

class Point {
    public int x;
    public int y;

    public Point(int a, int b) {
        x = a;
        y = b;
    }

    public String toString() {
        return "(" + x + ", " + y + ")";
    }
}